bug when server is shutting down (possibly after an error)

### Elasticsearch Version

docker.elastic.co/elasticsearch/elasticsearch:8.11.0

### Installed Plugins

N/A

### Java Version

_bundled_

### OS Version

MacOS Monterey 12.6

### Problem Description

Concurrent.ExecutionException","error.message":"java.lang.IllegalStateException: Can't move to stopped state when not started",

just look for "error" here:

```
elasticsearch-node3           | {"@timestamp":"2023-11-22T22:35:58.332Z", "log.level": "WARN", "message":"unexpected exception while waiting for http server to close", "ecs.version": "1.2.0","service.name":"ES_ECS","event.dataset":"elasticsearch.server","process.thread.name":"Thread-0","log.logger":"org.elasticsearch.node.Node","elasticsearch.node.name":"node3","elasticsearch.cluster.name":"mycluster","error.type":"java.util.concurrent.ExecutionException","error.message":"java.lang.IllegalStateException: Can't move to stopped state when not started","error.stack_trace":"java.util.concurrent.ExecutionException: java.lang.IllegalStateException: Can't move to stopped state when not started\n\tat java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)\n\tat java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)\n\tat org.elasticsearch.server@8.11.0/org.elasticsearch.node.Node.prepareForClose(Node.java:1776)\n\tat org.elasticsearch.server@8.11.0/org.elasticsearch.bootstrap.Elasticsearch.shutdown(Elasticsearch.java:468)\n\tat java.base/java.lang.Thread.run(Thread.java:1583)\nCaused by: java.lang.IllegalStateException: Can't move to stopped state when not started\n\tat org.elasticsearch.server@8.11.0/org.elasticsearch.common.component.Lifecycle.canMoveToStopped(Lifecycle.java:128)\n\tat org.elasticsearch.server@8.11.0/org.elasticsearch.common.component.AbstractLifecycleComponent.stop(AbstractLifecycleComponent.java:73)\n\tat org.elasticsearch.server@8.11.0/org.elasticsearch.node.Node.lambda$prepareForClose$59(Node.java:1768)\n\tat java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)\n\t... 1 more\n"}
elasticsearch-node1           | {"@timestamp":"2023-11-22T22:35:58.333Z", "log.level": "WARN", "message":"unexpected exception while waiting for http server to close", "ecs.version": "1.2.0","service.name":"ES_ECS","event.dataset":"elasticsearch.server","process.thread.name":"Thread-0","log.logger":"org.elasticsearch.node.Node","elasticsearch.node.name":"node1","elasticsearch.cluster.name":"mycluster","error.type":"java.util.concurrent.ExecutionException","error.message":"java.lang.IllegalStateException: Can't move to stopped state when not started","error.stack_trace":"java.util.concurrent.ExecutionException: java.lang.IllegalStateException: Can't move to stopped state when not started\n\tat java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)\n\tat java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)\n\tat org.elasticsearch.server@8.11.0/org.elasticsearch.node.Node.prepareForClose(Node.java:1776)\n\tat org.elasticsearch.server@8.11.0/org.elasticsearch.bootstrap.Elasticsearch.shutdown(Elasticsearch.java:468)\n\tat java.base/java.lang.Thread.run(Thread.java:1583)\nCaused by: java.lang.IllegalStateException: Can't move to stopped state when not started\n\tat org.elasticsearch.server@8.11.0/org.elasticsearch.common.component.Lifecycle.canMoveToStopped(Lifecycle.java:128)\n\tat org.elasticsearch.server@8.11.0/org.elasticsearch.common.component.AbstractLifecycleComponent.stop(AbstractLifecycleComponent.java:73)\n\tat org.elasticsearch.server@8.11.0/org.elasticsearch.node.Node.lambda$prepareForClose$59(Node.java:1768)\n\tat java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)\n\t... 1 more\n"}
elasticsearch-node3           | {"@timestamp":"2023-11-22T22:35:58.336Z", "log.level": "INFO", "message":"stopping ...", "ecs.version": "1.2.0","service.name":"ES_ECS","event.dataset":"elasticsearch.server","process.thread.name":"Thread-0","log.logger":"org.elasticsearch.node.Node","elasticsearch.node.name":"node3","elasticsearch.cluster.name":"mycluster"}
elasticsearch-node1           | {"@timestamp":"2023-11-22T22:35:58.337Z", "log.level": "INFO", "message":"stopping ...", "ecs.version": "1.2.0","service.name":"ES_ECS","event.dataset":"elasticsearch.server","process.thread.name":"Thread-0","log.logger":"org.elasticsearch.node.Node","elasticsearch.node.name":"node1","elasticsearch.cluster.name":"mycluster"}
elasticsearch-node3           | {"@timestamp":"2023-11-22T22:35:58.386Z", "log.level": "INFO", "message":"stopped", "ecs.version": "1.2.0","service.name":"ES_ECS","event.dataset":"elasticsearch.server","process.thread.name":"Thread-0","log.logger":"org.elasticsearch.node.Node","elasticsearch.node.name":"node3","elasticsearch.cluster.name":"mycluster"}
elasticsearch-node1           | {"@timestamp":"2023-11-22T22:35:58.387Z", "log.level": "INFO", "message":"stopped", "ecs.version": "1.2.0","service.name":"ES_ECS","event.dataset":"elasticsearch.server","process.thread.name":"Thread-0","log.logger":"org.elasticsearch.node.Node","elasticsearch.node.name":"node1","elasticsearch.cluster.name":"mycluster"}

```

### Steps to Reproduce

I have this in docker-compose:

```yaml
  elasticsearch-node1:
    image: "docker.elastic.co/elasticsearch/elasticsearch:8.11.0"
    container_name: elasticsearch-node1
    ulimits:
      memlock:
        soft: -1
        hard: -1
    environment:
      - cluster.name=mycluster
      - node.name=node1
      - action.auto_create_index=true
      - cluster.initial_master_nodes=node1  # Add node3
    ports:
      - "9201:9200"
```

### Logs (if relevant)

_No response_

--- a/docs/changelog/102759.yaml
+++ b/docs/changelog/102759.yaml
@@ -1,0 +1,6 @@
+pr: 102759
+summary: Close rather than stop `HttpServerTransport` on shutdown
+area: Infra/Node Lifecycle
+type: bug
+issues:
+ - 102501

--- a/server/src/main/java/org/elasticsearch/bootstrap/Elasticsearch.java
+++ b/server/src/main/java/org/elasticsearch/bootstrap/Elasticsearch.java
@@ -460,6 +460,8 @@ class Elasticsearch {
     }
 
     private static void shutdown() {
+        ElasticsearchProcess.markStopping();
+
         if (INSTANCE == null) {
             return; // never got far enough
         }

--- a/server/src/main/java/org/elasticsearch/bootstrap/ElasticsearchProcess.java
+++ b/server/src/main/java/org/elasticsearch/bootstrap/ElasticsearchProcess.java
@@ -1,0 +1,24 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.bootstrap;
+
+/**
+ * Helper class to determine if the ES process is shutting down
+ */
+public class ElasticsearchProcess {
+    private static volatile boolean stopping;
+
+    static void markStopping() {
+        stopping = true;
+    }
+
+    public static boolean isStopping() {
+        return stopping;
+    }
+}

--- a/server/src/main/java/org/elasticsearch/common/component/Lifecycle.java
+++ b/server/src/main/java/org/elasticsearch/common/component/Lifecycle.java
@@ -8,6 +8,8 @@
 
 package org.elasticsearch.common.component;
 
+import org.elasticsearch.bootstrap.ElasticsearchProcess;
+
 /**
  * Lifecycle state. Allows the following transitions:
  * <ul>
@@ -103,7 +105,11 @@ public final class Lifecycle {
             case STARTED -> false;
             case STOPPED -> {
                 assert false : "STOPPED -> STARTED";
-                throw new IllegalStateException("Can't move to started state when stopped");
+                throw new IllegalStateException(
+                    ElasticsearchProcess.isStopping()
+                        ? "Can't start lifecycle object when the Elasticsearch process is shutting down"
+                        : "Can't move to started state when stopped"
+                );
             }
             case CLOSED -> {
                 assert false : "CLOSED -> STARTED";

--- a/server/src/main/java/org/elasticsearch/node/Node.java
+++ b/server/src/main/java/org/elasticsearch/node/Node.java
@@ -575,10 +575,7 @@ public class Node implements Closeable {
      */
     public void prepareForClose() {
         HttpServerTransport httpServerTransport = injector.getInstance(HttpServerTransport.class);
-        FutureTask<Void> stopper = new FutureTask<>(() -> {
-            httpServerTransport.stop();
-            return null;
-        });
+        FutureTask<Void> stopper = new FutureTask<>(httpServerTransport::close, null);
         new Thread(stopper, "http-server-transport-stop").start();
 
         if (terminationHandler != null) {

--- a/server/src/main/java/org/elasticsearch/node/Node.java
+++ b/server/src/main/java/org/elasticsearch/node/Node.java
@@ -401,16 +401,7 @@ public class Node implements Closeable {
             }
         }
 
-        var httpTransport = injector.getInstance(HttpServerTransport.class);
-        try {
-            httpTransport.start();
-        } catch (IllegalStateException e) {
-            if (httpTransport.lifecycleState() == Lifecycle.State.CLOSED) {
-                throw new IllegalStateException("Node is shutting down, cannot complete startup", e);
-            } else {
-                throw e;
-            }
-        }
+        injector.getInstance(HttpServerTransport.class).start();
 
         if (WRITE_PORTS_FILE_SETTING.get(settings())) {
             TransportService transport = injector.getInstance(TransportService.class);

--- a/server/src/main/java/org/elasticsearch/node/Node.java
+++ b/server/src/main/java/org/elasticsearch/node/Node.java
@@ -401,7 +401,16 @@ public class Node implements Closeable {
             }
         }
 
-        injector.getInstance(HttpServerTransport.class).start();
+        var httpTransport = injector.getInstance(HttpServerTransport.class);
+        try {
+            httpTransport.start();
+        } catch (IllegalStateException e) {
+            if (httpTransport.lifecycleState() == Lifecycle.State.CLOSED) {
+                throw new IllegalStateException("Node is shutting down, cannot complete startup", e);
+            } else {
+                throw e;
+            }
+        }
 
         if (WRITE_PORTS_FILE_SETTING.get(settings())) {
             TransportService transport = injector.getInstance(TransportService.class);

--- a/server/src/test/java/org/elasticsearch/node/NodeTests.java
+++ b/server/src/test/java/org/elasticsearch/node/NodeTests.java
@@ -23,6 +23,7 @@ import org.elasticsearch.common.transport.BoundTransportAddress;
 import org.elasticsearch.core.RestApiVersion;
 import org.elasticsearch.env.Environment;
 import org.elasticsearch.gateway.PersistedClusterStateService;
+import org.elasticsearch.http.HttpServerTransport;
 import org.elasticsearch.index.IndexModule;
 import org.elasticsearch.index.IndexService;
 import org.elasticsearch.index.engine.Engine.Searcher;
@@ -78,6 +79,7 @@ import static org.hamcrest.Matchers.instanceOf;
 import static org.hamcrest.Matchers.is;
 import static org.hamcrest.Matchers.not;
 import static org.hamcrest.Matchers.nullValue;
+import static org.hamcrest.Matchers.startsWith;
 import static org.mockito.Mockito.mock;
 
 @LuceneTestCase.SuppressFileSystems(value = "ExtrasFS")
@@ -330,6 +332,13 @@ public class NodeTests extends ESTestCase {
         IllegalStateException e = expectThrows(IllegalStateException.class, () -> node.awaitClose(10L, TimeUnit.SECONDS));
         shard.store().decRef();
         assertThat(e.getMessage(), containsString("Something is leaking index readers or store references"));
+    }
+
+    public void testStartOnClosedTransport() throws IOException {
+        try (Node node = new MockNode(baseSettings().build(), basePlugins())) {
+            node.injector().getInstance(HttpServerTransport.class).close();
+            expectThrows(AssertionError.class, node::start);    // this would be IllegalStateException in a real Node with assertions off
+        }
     }
 
     public void testCreateWithCircuitBreakerPlugins() throws IOException {

--- a/server/src/test/java/org/elasticsearch/node/NodeTests.java
+++ b/server/src/test/java/org/elasticsearch/node/NodeTests.java
@@ -23,7 +23,6 @@ import org.elasticsearch.common.transport.BoundTransportAddress;
 import org.elasticsearch.core.RestApiVersion;
 import org.elasticsearch.env.Environment;
 import org.elasticsearch.gateway.PersistedClusterStateService;
-import org.elasticsearch.http.HttpServerTransport;
 import org.elasticsearch.index.IndexModule;
 import org.elasticsearch.index.IndexService;
 import org.elasticsearch.index.engine.Engine.Searcher;
@@ -79,7 +78,6 @@ import static org.hamcrest.Matchers.instanceOf;
 import static org.hamcrest.Matchers.is;
 import static org.hamcrest.Matchers.not;
 import static org.hamcrest.Matchers.nullValue;
-import static org.hamcrest.Matchers.startsWith;
 import static org.mockito.Mockito.mock;
 
 @LuceneTestCase.SuppressFileSystems(value = "ExtrasFS")
@@ -336,7 +334,7 @@ public class NodeTests extends ESTestCase {
 
     public void testStartOnClosedTransport() throws IOException {
         try (Node node = new MockNode(baseSettings().build(), basePlugins())) {
-            node.injector().getInstance(HttpServerTransport.class).close();
+            node.prepareForClose();
             expectThrows(AssertionError.class, node::start);    // this would be IllegalStateException in a real Node with assertions off
         }
     }

